### ISSUE TITLE
Scroll returns hits with scroll id before while(true)

### ISSUE THREAD

**Issue by Zelfapp** [elastic/elasticsearch-php#553]
### Scroll is returning hits on initial search, where I was expecting it to only return a [_scroll_id](https://www.elastic.co/guide/en/elasticsearch/client/php-api/5.0/_search_operations.html#_scan_scroll).

My size is set to 250. In `while(true)` the total hits expected are being returned, minus the scroll size of 250. So if there are 632 hits, the `while(true)` loop will only returning 382 _(632-250)_. 

The reason for this is because the `$docs =` part of script, which [the docs](https://www.elastic.co/guide/en/elasticsearch/client/php-api/5.0/_search_operations.html#_scan_scroll) state will only return a _scroll_id, is actually returning the scroll size of hits, before the while(true).

![returninghits](https://cloud.githubusercontent.com/assets/11559805/23666251/5941aede-030f-11e7-9d67-e68d3df4b836.jpg)

![hits](https://cloud.githubusercontent.com/assets/11559805/23666118/2914447e-030f-11e7-86af-f8903bb9f24f.jpg)

### Code snippet of problem

```php
$params = [
	'scroll' => '30s', // how long between scroll requests. should be small!
	'size' => 250, // how many results *per shard* you want back
	'index' => $this->index,
	'type' => $this->index,
	'body' => [
		'_source' => false,
		'query' => [
			'constant_score' => [
				'filter' => [
					'bool' => [
						'must' => [
							[
								'term' => [
									'tag_type' => 'dev_product'
								]
							],
							[
								'term' => [
									'user.exact' => 'some_user'
								]
							]
						]
					]
				]
			]
		]
	]
];

$docs = $client->search($params); // Execute the search

/******* The $docs contains hits, that will not be returned by scroll in while(true) loop ********/

$scrollId = $docs['_scroll_id']; // The response will contain no results, just a _scroll_id
$deleteIds['body'] = [];
$emailError = false;

while (true) { // Now we loop until the scroll "cursors" are exhausted
	$response = $client->scroll( // Execute a Scroll request
		[
			'scroll_id' => $scrollId, //...using previously obtained _scroll_id
			'scroll' => '30s' // and the same timeout window
		]
	);

	if (count($response['hits']['hits']) > 0) { // Check to see if any search hits from the scroll
		foreach($response['hits']['hits'] as $hit) {
			$deleteIds[] = $hit['_id'];
		}

		$scrollId = $response['_scroll_id']; // Get new scroll_id. Must always refresh _scroll_id!  It can change sometimes
	} else {
		break; // No results, scroll cursor is empty.  Exported all the data
	}
}
```

### System details

Running windows WAMP

- Windows 10
- PHP Version 5.6
- ES-PHP client version 5.1.3
- Elasticsearch version 5.0

**Comment by vladan-me** [comment]
Exactly this happened to me as well. I was using it normally and then upgraded to ES 5+ and suddenly the first scroll call is returning hits immediately. The issue is even worse if there's only one batch (all results are within scroll size), the second one is returning count but no hits. I am using similar configuration, PHP 5.6, ES-PHP 5.1.3 and Elasticsearch 5.2.1

**Comment by wavyx** [comment]
Same for me.
Does that mean the logic is now: search > while(consume > scroll) instead of search > while(scroll > consume) ?

**Comment by vladan-me** [comment]
@wavyx Basically, yes, assuming that you are executing everything continuously. I had a separate logic for retrieving scroll_id and then starting some kind of batch processing by rolling scroll_id's.
I believe that issue started by removing `search_type = scan`, since that is not an option in ES 5+, we are back to original scrolling mechanism (which returns hits immediately). That should be documented.

**Comment by wavyx** [comment]
@vladan-me indeed also hinted in https://github.com/elastic/elasticsearch/issues/16993#issuecomment-193667828
Also the doc from elasticsearch is not ideal and suggests only scroll id is returned while hits are send back too https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-scroll.html
(btw, the "scroll" has to be an URL parameter)

**Comment by polyfractal** [comment]
Urgh, this is my fault.  @vladan-me is correct, the documentation problem is caused by the removal of `scan`.  Previously you'd open a `scan` which contained nothing but the scroll ID, and then scroll over that.

When scan was removed, you just start scrolling so get results immediately.  I didn't even think of that when updating the client / docs.

Lemme fix up the docs right now, sorry!

**Comment by polyfractal** [comment]
Fixed in b951829.  Since both search and scroll now return `['hits']['hits']`, you should be able to still use a single loop.  I removed that `while(true)` though since it feels dangerous, better to just wait for the hits to come back empty.
